src/GitTools.jl

```diff
@@ -0,0 +1,127 @@
+module GitTools
+
+using ..Pkg3
+import LibGit2
+using Printf
+
+Base.@kwdef mutable struct MiniProgressBar
+    max::Float64 = 1
+    header::String = ""
+    color::Symbol = :white
+    width::Int = 40
+    current::Float64 = 0.0
+    prev::Float64 = 0.0
+    has_shown::Bool = false
+end
+
+function showprogress(io::IO, p::MiniProgressBar)
+    perc = p.current / p.max * 100
+    prev_perc = p.prev / p.max * 100
+    # Bail early if we are not updating the progress bar,
+    # Saves printing to the terminal
+    if p.has_shown && !(perc - prev_perc > 0.1)
+        return
+    end
+    p.prev = p.current
+    p.has_shown = true
+    n_filled = ceil(Int, p.width * perc / 100)
+    n_left = p.width - n_filled
+    print(io, "    ")
+    printstyled(io, p.header, color=p.color, bold=true)
+    print(io, " [")
+    print(io, "="^n_filled, ">")
+    print(io, " "^n_left, "]  ", )
+    @printf io "%2.1f %%" perc
+    print(io, "\r")
+end
+
+Base.@kwdef struct GitTransferProgress
+    total_objects::Cuint
+    indexed_objects::Cuint
+    received_objects::Cuint
+    local_objects::Cuint
+    total_deltas::Cuint
+    indexed_deltas::Cuint
+received_bytes::Csize_t
+end
+
+function transfer_progress(progress::Ptr{GitTransferProgress}, p::Any)
+
+    progress = unsafe_load(progress)
+    @assert p.transfer_progress != C_NULL
+    bar = unsafe_pointer_to_objref(p.transfer_progress)
```

Comment:
If you define this function as `transfer_progress(progress::Ptr{GitTransferProgress}, bar::MiniProgressBar)` you can perform the unsafe operations in another method like:
```julia
function transfer_progress(progress::Ptr{GitTransferProgress}, p::LibGit2.RemotePayloads)
    transfer_progress(progress, unsafe_pointer_to_objref(p.transfer_progress))
end
```

Reply:
What's the advantage?

Reply:
May make the code slightly cleaner